XenModel/Actions/OvfActions/Export.cs

```diff
@@ -0,0 +1,344 @@
+
+            int diskIndex = 0;
+            var vbdRefs = vm.VBDs;
+
+            for (int i = 0; i < vbdRefs.Count; i++)
+            {
+                int curVbd = i;
+                XenRef<VBD> vbdRef = vbdRefs[i];
+
+                VBD vbd = Connection.Resolve(vbdRef);
+                if (vbd == null)
+                    continue;
+
+                if (vbd.type == vbd_type.CD)
+                {
+                    string rasdid = OVF.AddCDROM(ovfEnv, vsId, vbd.uuid, OVF.GetContentMessage("RASD_16_CAPTION"), OVF.GetContentMessage("RASD_16_DESCRIPTION"));
+                    OVF.SetTargetDeviceInRASD(ovfEnv, vsId, rasdid, vbd.userdevice);
+                    continue;
+                }
+
+                VDI vdi = Connection.Resolve(vbd.VDI);
+                if (vdi == null)
+                    continue;
+
+                try
+                {
+                    var diskFilename = $"{vdi.uuid}.vhd";
+                    var diskPath = Path.Combine(targetDir, diskFilename);
+
+                    if (File.Exists(diskPath))
+                    {
+                        var oldFileName = diskFilename;
+                        diskFilename = $"{vdi.uuid}_{Thread.CurrentThread.ManagedThreadId}.vhd";
+                        diskPath = Path.Combine(targetDir, diskFilename);
+                        log.InfoFormat("VHD Name collision, renamed {0} to {1}", oldFileName, diskFilename);
+                    }
+
+                    string diskName = vdi.name_label;
+                    if (string.IsNullOrEmpty(diskName))
+                        diskName = $"{OVF.GetContentMessage("RASD_19_CAPTION")} {diskIndex}";
+
+                    if (!MetaDataOnly)
+                    {
+                        log.Info($"Exporting disk {diskName} to {diskFilename} for {vm.name_label}...");
+
+                        var taskRef = Task.create(Connection.Session, "export_raw_vdi_task",
+                            $"Exporting VDI {vdi.uuid} to {diskFilename}");
+
+                        HTTP_actions.get_export_raw_vdi(b =>
+                            {
+                                Description = string.Format(Messages.EXPORTING_VDI, diskName, diskFilename,
+                                    Util.DiskSizeString(b, 1, "F1"), Util.DiskSizeString(vdi.virtual_size));
+                                updatePercentage((curVbd + (float)b / vdi.virtual_size) / vbdRefs.Count);
+                            },
+                            () => Cancelling, XenAdminConfigManager.Provider.GetProxyTimeout(true),
+                            Connection.Hostname, XenAdminConfigManager.Provider.GetProxyFromSettings(Connection),
+                            diskPath, taskRef, Connection.Session.opaque_ref, vdi.uuid, "vhd");
+
+                        if (m_shouldVerify)
+                        {
+                            Description = string.Format(Messages.EXPORTING_VDI_VERIFICATION, diskFilename);
+
+                            //TODO

```

Comment:
I presume the verification is planned to be included in a later change?

Reply:
Yes, there is a separate ticket fo this.